vars/applier.groovy

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env groovy
+
+class ApplierInput implements Serializable{
+    String secretName = ""
+    String registryUrl = ""
+    String inventoryPath = ""
+    String requirementsPath = ""
+    String skipTlsVerify = false
+}
+
+def call(Map input) {
+    call(new ApplierInput(input))
+} 
+
+def call(ApplierInput input) {
+    assert input.secretName?.trim() : "Param secretName should be defined."
+    assert input.registryUrl?.trim() : "Param registryUrl should be defined."
+    assert input.inventoryPath?.trim() : "Param inventoryPath should be defined."
+    assert input.requirementsPath?.trim() : "Param requirementsPath should be defined."
+
+    script {
+        openshift.withCluster() {
+            def secretData = openshift.selector("secret/${input.secretName}").object().data
+            def encodedToken = secretData.token
+            env.TOKEN = sh(script:"set +x; echo ${encodedToken} | base64 --decode", returnStdout: true)
+        }
+
+        sh """
+            oc login ${input.registryUrl} --token=${env.TOKEN} --insecure-skip-tls-verify=${input.skipTlsVerify}
```

Comment:
Maybe i'm just a bit confused here. We're getting the secret _from_ openshift to then do an `oc login` _with_ openshift? this doesn't seem necessary. the applier script that i use doesn't need to do an `oc login`. if you can use `openshift.withCluster()` you should inherently be logged into your jenkins account.  

I guess this really begs the question, were we trying to log into a _different_ cluster than the one Jenkins is running on? if so, then why are we getting a secret directly from Openshift and not from Jenkins via `withCredentials`? Openshift will sync secrets here and then it gets away from having to use `openshift.withCluster()` making it usable when Jenkins isn't running on OCP

Also, FWIW Jenkins should be able to get it's own token via: `OCP_TOKEN = readFile('/var/run/secrets/kubernetes.io/serviceaccount/token').trim()`

Reply:
@pcarney8 the token is more for external clusters. i may have a dev, stage and prod cluster. once you go into the `sh`, you lose the scope of `openshift.withCluster()`. that's why i ended up just doing an `oc login` inside the `sh` to make sure i am running on the correct cluster.

Reply:
Hmm, ok. I'm still not huge on using `openshift.withCluster` to get your token then. You would then be storing access to another cluster in your cluster that has Jenkins. Which makes it seem like a security concern.

Also kind of confused why you would have a `dev` cluster Jenkins that you will then use to promote things to whole different cluster. 

Seems like it would make more sense to have your CD system either be _in_ that next cluster or completely outside of all the clusters. 

Either way. I'm all for optimistic merging, don't let my comments block this from getting merged. I'm not going to reject the PR

Reply:
@pcarney8 Let's say you have three clusters: DEV, STAGE and PROD (we could easily have more). We do not want the Jenkins master to be in each cluster. We really only want the Jenkins master to be on the DEV cluster and then build in DEV and promote throughout the environment. I am not sure how you have setup OpenShift clusters and promoted through the environment but happy to learn how you are looking at this.

Reply:
I'm on board with the Build Once philosophy, but deploying to PROD from a DEV Jenkins just feels wrong. I'm also not completely sold as Jenkins being an amazing CD tool, I think it's great for CI though. I think that might be where we differ in opinion. I don't want to hold this up more, feel free to direct message me and we can talk more.

I'm fine with the `oc login` piece and using the applier across clusters. 

Either way, the one **main** thing i would think would benefit from is this block:
```
        openshift.withCluster() {
            def secretData = openshift.selector("secret/${input.secretName}").object().data
            def encodedToken = secretData.token
            env.TOKEN = sh(script:"set +x; echo ${encodedToken} | base64 --decode", returnStdout: true)
        }
```
It assumes that Jenkins is running _on_ openshift. I think that's not a great assumption. If you use something like this:
```
node {
  withCredentials([string(credentialsId: 'mytoken', variable: 'TOKEN')]) {
    sh /* WRONG! */ """
      set +x
      curl -H 'Token: $TOKEN' https://some.api/
    """
    sh /* CORRECT */ '''
      set +x
      curl -H 'Token: $TOKEN' https://some.api/
    '''
  }
}
``` 
from: https://jenkins.io/doc/book/pipeline/jenkinsfile/#handling-credentials 
It gets away from that assumption and you can still get the secret that you need. All you need to know is that you are running _on_ Jenkins. The OCP sync plugin will handle the syncing of an openshift secret to your Jenkins (wherever it may live) and you don't have to worry about knowing that you are running _on_ Openshift. Your Jenkins could literally be running wherever and execute this `applier` function that you've made

Reply:
@pcarney8 We are getting more on the same page. Typically, you do use some CD tool to deploy to PROD. I agree being more generic where Jenkins is running as well. I feel like we should move forward with this as an initial step, but I would like to figure out a better solution in the next iteration. Also, I can't direct message you... I'm external :)

Reply:
@Gl4di4torRr ha, here I was thinking that Teams was already integrated with GH :stuck_out_tongue_closed_eyes: 

sounds good to me, go ahead and move forward with this and i'll add an issue to refactor like i mentioned above!